torch/utils/benchmark/visualize/table.py

```diff
@@ -0,0 +1,340 @@
+
+from torch.utils.benchmark.utils import common
+from torch.utils.benchmark.utils.valgrind_wrapper import timer_interface
+
+__all__ = ["Cell", "Layout", "Table"]
+
+
+T = TypeVar("T")
+REDUCTION_FUNCTION = Callable[[Tuple["Cell", ...]], T]
+REDUCTION_OUTPUT = Tuple[T, int]
+RESULT = Union[common.Measurement, timer_interface.CallgrindStats]
+RESULTS = Sequence[RESULT]
+
+
+# NB: We need to strip ANSI codes when computing width in order for
+#     columns to align properly when printed in terminal.
+# https://gist.github.com/rene-d/9e584a7dd2935d0f461904b9f2950007
+_ANSI_CODE: Pattern = re.compile(r"\033\[[0-9]+(;[0-9]+)?m")
+
+
+class Layout:
+    """Container for various style control enums."""
+
+    class HorizontalAlign(enum.Enum):
+        LEFT = 0
+        RIGHT = 1
+        CENTER = 2
+
+    class VerticalAlign(enum.Enum):
+        TOP = 0
+        BOTTOM = 1
+
+    ALIGNMENT_PAIR = Tuple[HorizontalAlign, VerticalAlign]
+
+    class ShouldRender(enum.Enum):
+        YES = 0
+        MAYBE = 1
+
+
+class Cell(abc.ABC):
+    """Core data container for `Table`. (And `Compare`)
+
+    One of the primary duties of `Compare` is to data layout. Columns need to be
+    aligned, empty cells need to be filled, row and column headers need to be
+    set, etc. This work is largely orthogonal to determining the contents of
+    the cells, except insofar as cell values might interact. (e.g. choosing a
+    common unit or aligning decimal places for readability.)
+
+    `Cell` attempts to separate these concerns by providing an interface where
+    an author can define arbitrary data processing cells while deferring the
+    more tedious and error-prone layout work to the framework.
+
+    For simple use cases, simply define a `render()` method which returns a
+    string representation of the cell (and optionally, override `alignment`
+    if a different alignment is desired), and the table will collect and render
+    the cells.
+
+    More complicated use cases will involve interaction between cells. For
+    instance, one needs to look at all entries of a row or column to color code
+    appropriately. For these cases, the `row_reduce` and `col_reduce` methods
+    can be used to implement cross cell data dependencies.
+    """
+
+    # TODO: Add `final` annotations once the minimum version is 3.8.
+    _row: Optional["Row"] = None
+
+    def __hash__(self) -> int:
+        return hash(id(self))
+
+    @abc.abstractmethod
+    def render(self) -> str:
+        ...
+
+    @property
+    def alignment(self) -> Layout.ALIGNMENT_PAIR:
+        return Layout.HorizontalAlign.RIGHT, Layout.VerticalAlign.TOP
+
+    @property
+    def should_render(self) -> Layout.ShouldRender:
+        return Layout.ShouldRender.YES
+
+    def set_context(self, row: "Row") -> None:
+        # We don't want to rely on __init__, as row is not known at construction.
+        self._row = row
+
+    def row_reduce(self, f: REDUCTION_FUNCTION) -> REDUCTION_OUTPUT:
```

Comment:
as these methods are expected to be used by end users in subclassed DataCells, they deserve to be documented